src/WithdrawalQueueManager.sol

```diff
@@ -0,0 +1,335 @@
+    //----------------------------------  WITHDRAWAL REQUESTS  -----------------------------
+    //--------------------------------------------------------------------------------------
+
+    function requestWithdrawal(uint256 amount) external nonReentrant {
+        if (amount <= 0) {
+            revert AmountMustBeGreaterThanZero();
+        }
+        
+        redeemableAsset.transferFrom(msg.sender, address(this), amount);
+
+        uint256 currentRate = redemptionAssetsVault.redemptionRate();
+        uint256 tokenId = _tokenIdCounter++;
+        withdrawalRequests[tokenId] = WithdrawalRequest({
+            amount: amount,
+            feeAtRequestTime: withdrawalFee,
+            redemptionRateAtRequestTime: currentRate,
+            creationTimestamp: block.timestamp,
+            creationBlock: block.number,
+            processed: false
+        });
+
+        pendingRequestedRedemptionAmount += calculateRedemptionAmount(amount, currentRate);
+
+        _mint(msg.sender, tokenId);
+
+        emit WithdrawalRequested(tokenId, msg.sender, amount);
+    }
+
+    //--------------------------------------------------------------------------------------
+    //----------------------------------  CLAIMS  ------------------------------------------
+    //--------------------------------------------------------------------------------------
+
+    function claimWithdrawal(uint256 tokenId, address receiver) public nonReentrant {
+        if (_ownerOf(tokenId) != msg.sender && _getApproved(tokenId) != msg.sender) {
+            revert CallerNotOwnerNorApproved(tokenId, msg.sender);
+        }
+
+        WithdrawalRequest memory request = withdrawalRequests[tokenId];
+
+        if (request.processed) {
+            revert WithdrawalAlreadyProcessed(tokenId);
+        }
+
+        if (!isFinalized(request)) {
+            revert NotFinalized(block.timestamp, request.creationTimestamp, secondsToFinalization);
+        }
+
+        withdrawalRequests[tokenId].processed = true;
+        pendingRequestedRedemptionAmount -= calculateRedemptionAmount(request.amount, request.redemptionRateAtRequestTime);
+
+        _burn(tokenId);
+        redeemableAsset.burn(request.amount);
+
+        uint256 unitOfAccountAmount = calculateRedemptionAmount(request.amount, request.redemptionRateAtRequestTime);
+
+        uint256 feeAmount = calculateFee(unitOfAccountAmount, request.feeAtRequestTime);
+        uint256 netUnitOfAccountAmount = unitOfAccountAmount - feeAmount;
+
+        uint256 currentBalance = redemptionAssetsVault.availableRedemptionAssets(address(this));
+        if (currentBalance < unitOfAccountAmount) {
+            revert InsufficientBalance(currentBalance, unitOfAccountAmount);
+        }
+
+        redemptionAssetsVault.transferRedemptionAssets(receiver, netUnitOfAccountAmount);
+        
+        if (feeAmount > 0) {
+            redemptionAssetsVault.transferRedemptionAssets(feeReceiver, feeAmount);
```

Comment:
At this point, we check three times that the balance is sufficient, is this necessary?